rust/datafusion/src/physical_plan/unicode_expressions.rs

```diff
@@ -0,0 +1,537 @@
+// the Postgres license.
+
+//! Unicode expressions
+
+use std::any::type_name;
+use std::cmp::Ordering;
+use std::str::from_utf8;
+use std::sync::Arc;
+
+use crate::error::{DataFusionError, Result};
+use arrow::{
+    array::{
+        ArrayRef, GenericStringArray, Int64Array, PrimitiveArray, StringOffsetSizeTrait,
+    },
+    datatypes::{ArrowNativeType, ArrowPrimitiveType},
+};
+use hashbrown::HashMap;
+use unicode_segmentation::UnicodeSegmentation;
+
+macro_rules! downcast_string_arg {
+    ($ARG:expr, $NAME:expr, $T:ident) => {{
+        $ARG.as_any()
+            .downcast_ref::<GenericStringArray<T>>()
+            .ok_or_else(|| {
+                DataFusionError::Internal(format!(
+                    "could not cast {} to {}",
+                    $NAME,
+                    type_name::<GenericStringArray<T>>()
+                ))
+            })?
+    }};
+}
+
+macro_rules! downcast_arg {
+    ($ARG:expr, $NAME:expr, $ARRAY_TYPE:ident) => {{
+        $ARG.as_any().downcast_ref::<$ARRAY_TYPE>().ok_or_else(|| {
+            DataFusionError::Internal(format!(
+                "could not cast {} to {}",
+                $NAME,
+                type_name::<$ARRAY_TYPE>()
+            ))
+        })?
+    }};
+}
+
+/// Returns number of characters in the string.
+/// character_length('josé') = 4
+pub fn character_length<T: ArrowPrimitiveType>(args: &[ArrayRef]) -> Result<ArrayRef>
+where
+    T::Native: StringOffsetSizeTrait,
+{
+    let string_array: &GenericStringArray<T::Native> = args[0]
+        .as_any()
+        .downcast_ref::<GenericStringArray<T::Native>>()
+        .ok_or_else(|| {
+            DataFusionError::Internal("could not cast string to StringArray".to_string())
+        })?;
+
+    let result = string_array
+        .iter()
+        .map(|string| {
+            string.map(|string: &str| {
+                T::Native::from_usize(string.graphemes(true).count()).unwrap()
+            })
+        })
+        .collect::<PrimitiveArray<T>>();
+
+    Ok(Arc::new(result) as ArrayRef)
+}
+
+/// Returns first n characters in the string, or when n is negative, returns all but last |n| characters.
+/// left('abcde', 2) = 'ab'
+pub fn left<T: StringOffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    let string_array = downcast_string_arg!(args[0], "string", T);
+    let n_array = downcast_arg!(args[1], "n", Int64Array);
+
+    let result = string_array
+        .iter()
+        .zip(n_array.iter())
+        .map(|(string, n)| match (string, n) {
+            (None, _) => None,
+            (_, None) => None,
+            (Some(string), Some(n)) => match n.cmp(&0) {
+                Ordering::Equal => Some(""),
+                Ordering::Greater => Some(
+                    string
+                        .grapheme_indices(true)
+                        .nth(n as usize)
+                        .map_or(string, |(i, _)| {
+                            &from_utf8(&string.as_bytes()[..i]).unwrap()
```

Comment:
For any `unwrap()` I think it makes sense to put a comment why the unwrap is ok, and/or use `.expect`

Reply:
I spent some time poking around in https://docs.rs/unicode-segmentation/1.7.1/unicode_segmentation/trait.UnicodeSegmentation.html#tymethod.graphemes

I couldn't find anything that would let you get back a `&str` from the grapheme indices. I wonder if it might be possible to use the `graphemes()` call itself and count up the lengths of the `&str` that came back

https://docs.rs/unicode-segmentation/1.7.1/unicode_segmentation/trait.UnicodeSegmentation.html#tymethod.graphemes

Something like (untested)

```rust
let i = strings
  .graphemes()
  .limit(n as usize)
  .map(|s| s.len())
  .sum()

string[..i]
```

Or something like that. I am not sure how important this is

Reply:
Let me have a look and see if we can do this.

Reply:
@alamb I have refactored just the `left` function to demonstrate how it could be done:

https://github.com/apache/arrow/blob/7e40d25c8980102de48b26b88c04258b7fa04109/rust/datafusion/src/physical_plan/unicode_expressions.rs#L98

I think this is nicer so if you agree I can apply to the rest of the code?

Reply:
I like it a lot 👍 Thanks @seddonm1 

I am not sure why this `clone` is needed, but that is a minor point and can be cleaned up later
```
                    let len = graphemes.clone().count() as i64;
```

Reply:
The borrow checker is identifying that the second use of `graphemes` is using a moved value without the clone. If you can help resolve that it would help me learn.

```
use of moved value: `graphemes` value used here after move
```

Reply:
I see -- I took a closer look and I see now that `graphemes` is an iterator -- so we are `clone()` ing an iterator, which I guess feels right to me -- the code has to effectively scan the input to figure out how many graphemes long it is and then scan it again to create the output. 

Sorry for my confusion!